src/pdn/src/connect.h

```diff
@@ -0,0 +1,163 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include <map>
+#include <set>
+#include <vector>
+
+#include "shape.h"
+
+namespace odb {
+class dbVia;
+class dbTechViaGenerateRule;
+class dbTechLayer;
+class dbTechVia;
+class dbSWire;
+class Rect;
+}  // namespace odb
+
+namespace pdn {
+
+class DbVia;
+class ViaGenerator;
+
+class Connect
+{
+ public:
+  Connect(Grid* grid, odb::dbTechLayer* layer0, odb::dbTechLayer* layer1);
+
+  void addFixedVia(odb::dbTechViaGenerateRule* via);
+  void addFixedVia(odb::dbTechVia* via);
+
+  void setCutPitch(int x, int y);
+  int getCutPitchX() const { return cut_pitch_x_; }
+  int getCutPitchY() const { return cut_pitch_y_; }
+
+  void setMaxRows(int rows) { max_rows_ = rows; }
+  int getMaxRows() const { return max_rows_; }
+
+  void setMaxColumns(int cols) { max_columns_ = cols; }
+  int getMaxColumns() const { return max_columns_; }
+
+  void setOnGrid(const std::vector<odb::dbTechLayer*>& layers);
+
+  void setSplitCuts(const std::map<odb::dbTechLayer*, int>& splits);
+  int getSplitCutPitch(odb::dbTechLayer* layer) const;
+
+  void report() const;
+
+  odb::dbTechLayer* getLowerLayer() const { return layer0_; }
+  odb::dbTechLayer* getUpperLayer() const { return layer1_; }
+
+  bool isSingleLayerVia() const;
+  bool isMultiLayerVia() const { return !isSingleLayerVia(); }
+  bool isTaperedVia(const odb::Rect& lower, const odb::Rect& upper) const;
+
+  bool hasCutPitch() const { return cut_pitch_x_ != 0 || cut_pitch_y_ != 0; }
+
+  const std::vector<odb::dbTechLayer*>& getIntermediteLayers() const
+  {
+    return intermediate_layers_;
+  }
+  const std::vector<odb::dbTechLayer*>& getIntermediteRoutingLayers() const
+  {
+    return intermediate_routing_layers_;
+  }
+  std::vector<odb::dbTechLayer*> getAllLayers() const;
+  std::vector<odb::dbTechLayer*> getAllRoutingLayers() const;
+  bool containsIntermediateLayer(odb::dbTechLayer* layer) const;
+  bool overlaps(const Connect* connect) const;
+  bool startsBelow(const Connect* connect) const;
+
+  bool appliesToVia(const ViaPtr& via) const;
+
+  void makeVia(odb::dbSWire* wire,
+               const odb::Rect& lower_rect,
+               const odb::Rect& upper_rect,
+               odb::dbWireShapeType type,
+               DbVia::ViaLayerShape& via_shapes);
+
+  void setGrid(Grid* grid) { grid_ = grid; }
+  Grid* getGrid() const { return grid_; }
+
+  void clearShapes();
+
+  void filterVias(const std::string& filter);
+
+ private:
+  Grid* grid_;
+  odb::dbTechLayer* layer0_;
+  odb::dbTechLayer* layer1_;
+  std::vector<odb::dbTechViaGenerateRule*> fixed_generate_vias_;
+  std::vector<odb::dbTechVia*> fixed_tech_vias_;
+  int cut_pitch_x_;
+  int cut_pitch_y_;
+
+  int max_rows_;
+  int max_columns_;
+
+  std::set<odb::dbTechLayer*> ongrid_;
+  std::map<odb::dbTechLayer*, int> split_cuts_;
+
+  std::map<std::pair<int, int>, std::unique_ptr<DbVia>> vias_;
```

Comment:
Comment what the int pair represetnts

Reply:
I'll add the comment, but it represents the height and width of a given set of vias. So if I've already created a stack for a particular size, I can just use that instead of rebuilding.